### ISSUE TITLE
Vulkan application crashes immediately when running through RenderDoc

### ISSUE THREAD

**Issue by RedNicStone** [baldurk/renderdoc#2486]
## Description
I've been running into a problem lately where my application would crash on startup when trying to capture with RenderDoc.
This issue seems to be hardware / setup specific as it runs just fine on my notebook.

## Steps to reproduce
This is the application I am trying to capture from: https://github.com/RedNicStone/CubiCAD
Unfortunately I was unable to reproduce this issue on my notebook.
These are the last log entries RenderDoc prints before the application (CubiCAD) crashes:
```
Core     PID  41317: [22:59:07]              core.cpp(480) - Log     - RenderDoc v1.17 Linux 64-bit Release (NO_GIT_COMMIT_HASH_DEFINED) loaded in replay application
Core     PID  41317: [22:59:07]              core.cpp(494) - Warning - Couldn't open socket for target control
Core     PID  41317: [22:59:07]          settings.cpp(460) - Log     - Loading config from /home/nic/.renderdoc/renderdoc.conf
UI       PID  41317: [22:59:07]                        (0) - Log     - QRenderDoc initialising.
UI       PID  41317: [22:59:07]                        (0) - Log     - Disabling error reports due to invalid commit hash
Core     PID  41317: [22:59:07]     android_tools.cpp(347) - Log     - Initialising adb using 'adb'
Core     PID  41317: [22:59:07]     android_tools.cpp(352) - Warning - Couldn't locate adb. Ensure adb is in PATH, ANDROID_SDK or ANDROID_HOME is set, or you configure your SDK location
Core     PID  41317: [22:59:07]        vk_replay.cpp(4498) - Log     - Created device.
UI       PID  41317: [22:59:07]                        (0) - Warning - area parameter ignored for this type
Core     PID  41317: [22:59:15]     posix_process.cpp(951) - Log     - Running process /home/nic/CLionProjects/CubiCAD/CubiCAD for injection
Core     PID  41349: [22:59:16]              core.cpp(480) - Log     - RenderDoc v1.17 Linux 64-bit Release (NO_GIT_COMMIT_HASH_DEFINED) capturing application
Core     PID  41349: [22:59:16]          settings.cpp(460) - Log     - Loading config from /home/nic/.renderdoc/renderdoc.conf
Core     PID  41349: [22:59:16]     posix_libentry.cpp(60) - Log     - Using delay for debugger 0
Core     PID  41349: [22:59:16]     posix_libentry.cpp(73) - Log     - Loading into /home/nic/CLionProjects/CubiCAD/CubiCAD
Core     PID  41349: [22:59:16]          gl_hooks.cpp(263) - Log     - Registering OpenGL hooks
Core     PID  41349: [22:59:16]         glx_hooks.cpp(793) - Log     - Registering GLX hooks
Core     PID  41349: [22:59:16]         egl_hooks.cpp(992) - Log     - Registering EGL hooks
Core     PID  41349: [22:59:16]           vk_layer.cpp(99) - Log     - Registering Vulkan hooks
Core     PID  41317: [22:59:16]    target_control.cpp(587) - Log     - Got remote handshake: CubiCAD [41349]
Core     PID  41349: [22:59:16]             core.cpp(1845) - Log     - Adding Vulkan device frame capturer for 0x00000000020ADFB0
Core     PID  41349: [22:59:16]   vk_device_funcs.cpp(847) - Log     - Initialised capture layer in Vulkan instance.
Core     PID  41349: [22:59:16]  vk_device_funcs.cpp(1461) - Log     - physical device 0: llvmpipe (LLVM 13.0.0, 256 bits) (ver 0.0 patch 0x1) - 10005:0000
Core     PID  41349: [22:59:16]  vk_device_funcs.cpp(1461) - Log     - physical device 1: NVIDIA GeForce RTX 2080 Ti (ver 510.39 patch 0x100) - 10de:1e04
Core     PID  41349: [22:59:16]             core.cpp(1877) - Log     - Adding Vulkan frame capturer for 0x00000000020ADFB0 / 0x0000000005C00007
Core     PID  41349: [22:59:16]          vk_core.cpp(1621) - Warning - VkPhysicalDeviceBufferDeviceAddressFeaturesKHR.bufferDeviceAddressCaptureReplay is false, can't support capture of VK_KHR_buffer_device_address
```
I tried collecting call stacks but I was unable to do so due to some strange behaviour on the side of renderdoccmd (It keeps changing directories to ~/OpenFOAM/nic-9/run for some godforsaken reason)

## Environment

* RenderDoc version: qtrenderdoc / renderdoccmd x64 v1.17
* Operating System: fedora 35 running v5.16.5-200 GNU/Linux on x86/64
* Graphics API: Vulkan 1.2.189
* DRM Driver: nV drv v510.39.01
* CUDA Compiler: v11.6
* Graphics Hardware: nVidia RTX 2080ti

**Comment by baldurk** [comment]
I tried to run this project to reproduce but I can't. On windows it doesn't build for me but I'm not sure if it's supposed to work. On linux it builds but fails to run. Full log is [here](https://github.com/baldurk/renderdoc/files/8019607/log.txt), this is the final output:

```
Graphics queue: 0x55c634fc3e40
Transfer queue: 0x55c634fc3e40
Compute queue: 0x55c634fc3e40
Present queue: 0x55c634fc3e40
validation layer: Validation Error: [ VUID-VkImageCreateInfo-extent-02253 ] Object 0: handle = 0x55c634eb0ab0, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0x483d7c67 | vkCreateImage(): extent.height 880469472 exceeds allowable maximum image extent height 32768. The Vulkan spec states: extent.height must be less than or equal to imageCreateMaxExtent.height (as defined in Image Creation Limits) (https://www.khronos.org/registry/vulkan/specs/1.3-extensions/html/vkspec.html#VUID-VkImageCreateInfo-extent-02253)
validation layer: Validation Error: [ VUID-vkAllocateMemory-pAllocateInfo-01713 ] Object 0: handle = 0x55c634eb0ab0, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0xe9a2b96f | vkAllocateMemory: attempting to allocate 77368616255488 bytes from heap 0,but size of that heap is only 11811160064 bytes. The Vulkan spec states: pAllocateInfo->allocationSize must be less than or equal to VkPhysicalDeviceMemoryProperties::memoryHeaps[memindex].size where memindex = VkPhysicalDeviceMemoryProperties::memoryTypes[pAllocateInfo->memoryTypeIndex].heapIndex as returned by vkGetPhysicalDeviceMemoryProperties for the VkPhysicalDevice that device was created from (https://www.khronos.org/registry/vulkan/specs/1.3-extensions/html/vkspec.html#VUID-vkAllocateMemory-pAllocateInfo-01713)
validation layer: Validation Error: [ VUID-vkAllocateMemory-pAllocateInfo-01713 ] Object 0: handle = 0x55c634eb0ab0, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0xe9a2b96f | vkAllocateMemory: attempting to allocate 77368616255488 bytes from heap 1,but size of that heap is only 12538767360 bytes. The Vulkan spec states: pAllocateInfo->allocationSize must be less than or equal to VkPhysicalDeviceMemoryProperties::memoryHeaps[memindex].size where memindex = VkPhysicalDeviceMemoryProperties::memoryTypes[pAllocateInfo->memoryTypeIndex].heapIndex as returned by vkGetPhysicalDeviceMemoryProperties for the VkPhysicalDevice that device was created from (https://www.khronos.org/registry/vulkan/specs/1.3-extensions/html/vkspec.html#VUID-vkAllocateMemory-pAllocateInfo-01713)
could not allocate memory
```

You should run your application with the vulkan validation layers and be sure you fix any issues identified, as issues like this with undefined behaviour can cause crashes under RenderDoc.

If you can fix this issue and let me run the program I can look into it further but I can't do anything unless I can run the repro case.

Also unfortunately I've found it unreliable trying to attach a debugger to running processes in RenderDoc. In principle you should be able to set 'delay for debugger' to e.g. 30 seconds and attach a debugger to the process and debug from the start, but I've found gdb either conflicts or gets into a broken state where it doesn't properly continue and instead keeps stopping the process and doesn't properly continue.

**Comment by RedNicStone** [comment]
Yes, the application is only designed to run on linux operating systems. Thanks for taking the time and trying to run it, I very much appreciate that!
It seems like it fails to load textures which probably is due to borked dependencies on the repository side (At the time you ran the program there was no check if a file actually existed). I've fixed the issue with the newest commit and it should work now.
The program currently fails to render anything, which is one of the reasons I am trying to debug it with RenderDoc. The issue of RenderDoc not being able to run the application existed beforehand (It did manage to run at one point but loading the capture would result in RenderDoc throwing an error about incompatible hardware).
If the application is built in debug mode (which i suggest doing) validation layers will be used. I do sometimes get errors when closing the window but its inconsistent and dosent seem to be the cause of why the program crashes when run inside RenderDoc.

**Comment by baldurk** [comment]
I can now run the program, but unfortunately there's no problem for me and I'm able to run it and capture successfully. I do see the black screen - the rendering seems fine and RenderDoc's text overlay doesn't appear so I'm guessing it's an issue with surface handling or presentation. I guess if this only happens on some hardware I'm not able to reproduce it on mine, though it's the same GPU and as similar driver as I could reasonably get with distro/linux issues (510.47). I also tried on my AMD system but got the same results.

You could try to use gdb but it's quite brittle. I was able to attach using a command like `gdb -ex 'handle SIGSTOP ignore' -ex 'b Instance::create' ./CubiCAD $(pgrep CubiCAD)` after launching the program with a 'debugger delay' set to 30 seconds. You may find RenderDoc doesn't connect properly automatically due to timeout issues, and you'll still have to `cont` gdb a few times to wake each thread up, but it might help you get to the crash. I don't know why you need to ignore SIGSTOP signals, but if you don't then gdb keeps stopping on them forever - I don't think ptrace is designed to be used by multiple processes.

The other alternative is to use printf or logging statements to narrow down where the crash is happening. If you can give me more information about what is crashing and when I can try to help, but unfortunately unless I'm able to reproduce I can't suggest much.

**Comment by RedNicStone** [comment]
I found a rather simple way to attach the debugger to a renderdoc process:
1. In your `int main()` function place a `getchar();` at the top
2. Run the program trough renderdoccpi and copy the PID that renderdoc prints out
3. Connect your debugger to that PID and start debugging :factory_worker:
Hope this also helps with future issues

Using this method I was able to collect call stacks as well as all other kinds of meta data.
The crash is caused by a segmentation fault (yay my favourite!) due to a function call to nullptr.
Here is the call stack just before the fault:
```
<unknown> 0x0000000000000000
vkBasalt::vkBasalt_CreateDevice(VkPhysicalDevice_T*, VkDeviceCreateInfo const*, VkAllocationCallbacks const*, VkDevice_T**) 0x00007f81d4ba4f55
loader_create_device_chain 0x00007f81f6cb75e6
loader_layer_create_device 0x00007f81f6cbddb1
vkCreateDevice 0x00007f81f6ccee93
Device::create device.cpp:66
RenderManager::createLogicalDevice rendermanager.cpp:54
RenderManager::createRenderObjects rendermanager.cpp:231
RenderManager::create rendermanager.cpp:21
VulkanRasterizer::createRenderManager vulkan_voxel_rasterizer_4.cpp:96
VulkanRasterizer::init vulkan_voxel_rasterizer_4.cpp:102
VulkanRasterizer::run vulkan_voxel_rasterizer_4.cpp:62
main vulkan_voxel_rasterizer_4.cpp:171
__libc_start_call_main 0x00007f81f582d560
__libc_start_main_impl 0x00007f81f582d60c
_start 0x000000000040a8d5
```
As can be seen the issue does not originate from my application but from ... somewhere completely different (Which might be something to look into....).
vkBasalt is a completely unrelated application that is not (directly) used by my application. Its a software somewhat similar to reShade but for vulkan on linux. [Here is their git](https://github.com/DadSchoorse/vkBasalt). I never installed vkBasalt directly but it may have been installed together with a GoOverlay some time ago. Apparently it is enabled globally: 
```
[nic@ryzen9m CubiCAD]$ echo "$ENABLE_VKBASALT"
1
```
This might be because of broken linkage between vkBasalt and vulkan but I'm not sure.
Setting `ENABLE_VKBASALT` to 0 to disable it did not fix the issue however as vkGetDeviceQueue is now causing issues:
```
vkSetDeviceDispatch 0x00007fbe09cdf4bf
WrappedVulkan::vkGetDeviceQueue(VkDevice_T*, unsigned int, unsigned int, VkQueue_T**) 0x00007fbe0a593409
vkGetDeviceQueue 0x00007fbe09cee9d3
QueueFamily::getQueues queues.cpp:206
QueueFamilyHandler::getQueues queues.cpp:144
Device::create device.cpp:70
RenderManager::createLogicalDevice rendermanager.cpp:54
RenderManager::createRenderObjects rendermanager.cpp:231
RenderManager::create rendermanager.cpp:21
VulkanRasterizer::createRenderManager vulkan_voxel_rasterizer_4.cpp:96
VulkanRasterizer::init vulkan_voxel_rasterizer_4.cpp:102
VulkanRasterizer::run vulkan_voxel_rasterizer_4.cpp:62
main vulkan_voxel_rasterizer_4.cpp:171
__libc_start_call_main 0x00007fbe08852560
__libc_start_main_impl 0x00007fbe0885260c
_start 0x000000000040a8d5
```
This causes another segmentation fault in vulkan itself this time. The program seems to fail when requesting the handle to and other than the first queue from queueFamily with ID 0. However there was a quite sus console output: `WARNING: lavapipe is not a conformant vulkan implementation, testing use only.`. I don't know why this happens exactly (The application typically picks non-lavapipe implementations) but I manually disabled picking lavapipe as implementation and that did the trick!

![Screenshot from 2022-02-09 21-48-48](https://user-images.githubusercontent.com/36967667/153296340-9f06d56b-456d-4c53-bbe0-2a04e14ffc72.png)
![Screenshot from 2022-02-09 21-53-12](https://user-images.githubusercontent.com/36967667/153296428-1f8e28c8-f769-43a1-90a6-7be25e39739c.png)

Working even better than before now :+1: :100: :top:. Where is previously had some issues with "Incompatible replay hardware" its running now. Thanks so much for taking the time and effort to help resolve this issue! I probably wouldn't even be able to attach a debugger still :/.

EDIT: Should this issue be considered closed? Maybe setting `ENABLE_VKBASALT` to 0 if not manually set to 1 could prevent compatibility issues...
